### ISSUE TITLE
Problem with multiple accounts

### ISSUE THREAD

**Issue by emreoktem** [p2/OAuth2#111]
Hi. I'm using this framework in my swift based OSX app and it works great.

In our project, we need to allow user to connect to the same service by using two different accounts. 

In order to login to the first service, we use the new embedded view to show the login service and everything works ok. 

When the user wants to connect to the same service with his second account the embedded view doesn't show the login window because the user has already logged in. The embedded view automatically calls the user back and dismissed. This is the part we found a workaround by deleting cookies of the OAuth2WebViewController during the second login.

On the other hand, because the tokens are saved to keychain, even the second login is successful, the authorize function cannot use the second keychain element and the app cannot access by using the second login.

Support for two logins to the same service or a work around will be appreciated.

Cheers...

**Comment by p2** [comment]
That's a good use-case. I'm trying to think how you would use the library in this case – would you have one `OAuth2` instance per account or should `authorize()` take an optional parameter for which account to use? I'd choose the former route since most of the time you don't care about accounts with OAuth2, but if you can make an argument for the latter form I'm interested to hear it.

**Comment by emreoktem** [comment]
Hi,

Thanks for the feedback. 

I have a custom class named as service and each service class creates it's own Oauth2 instance with following configuration.

` var oauth2 : OAuth2CodeGrant = OAuth2CodeGrant(settings: [
        "client_id": "xxxxx",
        "client_secret": "xxxxx",
        "authorize_uri": "https://github.com/login/oauth/authorize",
        "token_uri": "https://github.com/login/oauth/access_token",
        "scope": "user,repo",
        "redirect_uris": ["myOauth://oauth/callback/github"], 
        "secret_in_body": true,
        "keychain":true,
        ])`

and here is the configuration of Oauth2

`Oauth2.authConfig.authorizeEmbedded = true`
`Oauth2.authConfig.authorizeContext = windowController`
`Oauth2.authorizeEmbeddedFrom(windowController)`

And I call the above authorization method separately for each of my services.

As mentioned before, If I have two instances of the same service (for example, Github) the first login is ok but the second one uses the credentials of the first.

Again thanks.

**Comment by p2** [comment]
How do you tell OAuth2 which account to use?

**Comment by p2** [comment]
You can now change `keychainAccountForTokens` on each _OAuth2_ instance, but I'm not sure if that is useful to you (since you don't know the account the user is using, which is by design in the OAuth2 protocol).

**Comment by emreoktem** [comment]
I am currently creating two different OAuth2CodeGrant object instances with the exact same parameters because I need to connect to the same service but by using two different accounts. Let me try if the new method works or not. Thanks for the support.

**Comment by vojto** [comment]
Hi,

thanks for all the pointers in this thread, I found this quite easy - and the framework is ready to be used with multiple accounts, IMO.

Some things I discovered:

- Keep track of your accounts. Have a table with account alias (user-entered, so they could make sense of them) and the identifier.
- You generate the identifier for each account. Generate the identifier, set it as `keychainAccountForTokens`, let user login - and once login goes through, save that identifier with the account.
- All subsequent requests for that account, you'll create a new instance of Oauth and set your identifier as `keychainAccountForTokens`.
- Reset (remove cookies) the webview each time before logging in. ([This code](https://github.com/p2/OAuth2/pull/162#issuecomment-266029053) contains some hints - wip)

It would be nice to also download user's email and associate it with the Account, so we could show it in the UI. 

Btw. using this library for [this project](https://median.tech), and so far it's been a breeze working with it. Great job @p2!
